samples: Bluetooth: hci_uart: Implement NOP Command Complete

carlescufi opened the pull request:
Implement the Controller counterpart to CONFIG_BT_WAIT_NOP so that it
issues a NOP Command Complete event after booting up, to signal to the
Host that it is ready to receive HCI traffic.

Fixes #15333

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
samples: Bluetooth: hci_uart: Implement NOP Command Complete

Implement the Controller counterpart to CONFIG_BT_WAIT_NOP so that it
issues a NOP Command Complete event after booting up, to signal to the
Host that it is ready to receive HCI traffic.

Fixes #15333

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
Output of a reset with this patch:

![image](https://user-images.githubusercontent.com/12450381/55888363-8331e380-5baf-11e9-93ba-9a335683e0d4.png)

lemrey commented:
This should address #15333.

jhedberg commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -357,6 +357,32 @@ void main(void)
+			},
+			.cc = {
+				.ncmd = 1,
+				.opcode = BT_OP_NOP,
```
I know that it makes no difference in practice, however this should really be sys_cpu_to_le16(BT_OP_NOP) to keep track of what endianness `.opcode` is.

carlescufi replied:
Is this actually allowed on an initializer?

jhedberg replied:
On big endian this maps to `__bswap_16()` which is defined as follows:

```
#define __bswap_16(x) ((u16_t) ((((x) >> 8) & 0xff) | (((x) & 0xff) << 8))) 
```

I'd think the compiler should be able to handle that even for the initialiser of a const variable, but you could always try it by using `__bswap_16(BT_OP_NOP)`

carlescufi replied:
It is. Updated, thanks!

jhedberg reviewed (changes requested):
Looks generally good, except for one minor endianness "correctness" issue

carlescufi committed:
samples: hci_uart: Reduce MAX_CONN to 16 for the micro:bit

In order to fit in SRAM, reduce the CONFIG_BT_MAX_CONN from 20 to 16 in
the configuration used for the BBC micro:bit.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

anangl commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -357,6 +357,32 @@ void main(void)
+		};
+
+		for (i = 0; i < sizeof(cc_evt); i++) {
+			uart_poll_out(hci_uart_dev, *(((u8_t *)&cc_evt)+i));
```
`(const u8_t *)`, just for consistency

carlescufi replied:
Fixed, thank you

lemrey reviewed (approved):
LGTM
